modules/imgproc/src/subdivision2d.cpp

```diff
@@ -210,19 +210,126 @@ void Subdiv2D::swapEdges( int edge )
     splice(sedge, getEdge(b, NEXT_AROUND_LEFT));
 }
 
-static double triangleArea( Point2f a, Point2f b, Point2f c )
-{
-    return ((double)b.x - a.x) * ((double)c.y - a.y) - ((double)b.y - a.y) * ((double)c.x - a.x);
+static double CrossProduct(Point2f u, Point2f v) {
+    return (double)u.x * v.y - (double)v.x * u.y;
 }
 
-int Subdiv2D::isRightOf(Point2f pt, int edge) const
-{
-    Point2f org, dst;
-    edgeOrg(edge, &org);
-    edgeDst(edge, &dst);
-    double cw_area = triangleArea( pt, dst, org );
+static double DotProduct(Point2f u, Point2f v) {
+    return (double)u.x * v.x + (double)u.y * v.y;
+}
+
+static double Distance(Point2f a, Point2f b) {
+    return sqrt(DotProduct(b - a, b - a));
+}
+
+static bool Collinear(Point2f u, Point2f v) {
+    return abs(CrossProduct(u, v)) < FLT_EPSILON;
+}
+
+static bool InRectangle(Point2f x, Point2f a, Point2f b) {
+    return  min(a.x, b.x) < x.x && x.x < max(a.x, b.x) &&
+            min(a.y, b.y) < x.y && x.y < max(a.y, b.y);
+}
 
-    return (cw_area > 0) - (cw_area < 0);
+static int CCW(Point2f a, Point2f b, Point2f c) {
+    double cp = CrossProduct(b - a, c - a);
+    return cp > FLT_EPSILON ? 1 : (cp < -FLT_EPSILON ? -1 : 0);
+}
+
+static int LeftOf(Point2f x, Point2f a, Point2f b) {
+    return CCW(x, a, b);
+}
+
+static int InCircle(Point2f a, Point2f b, Point2f c, Point2f d) {
+    const double eps = FLT_EPSILON * 0.125;
+
+    double val =
+            ((double)a.x * a.x + (double)a.y * a.y) * CrossProduct( c - b, d - b );
+    val -=  ((double)b.x * b.x + (double)b.y * b.y) * CrossProduct( c - a, d - a );
+    val +=  ((double)c.x * c.x + (double)c.y * c.y) * CrossProduct( b - a, d - a );
+    val -=  ((double)d.x * d.x + (double)d.y * d.y) * CrossProduct( b - a, c - a );
+
+    return val > eps ? 1 : val < -eps ? -1 : 0;
+}
+
+#define META(i) ((i) == 1 || (i) == 2 || (i) == 3)
+const Point2f ORIGIN(0.f, 0.f);
+
+int Subdiv2D::CCWEx(int i, int j, int k) const {
+
+    if (META(i) && META(j) && META(k)) {
+        return CCW(vtx[i].pt, vtx[j].pt, vtx[k].pt);
+    }
+
+    if (META(i) && META(j) && !META(k)) {
+        return CCW(vtx[i].pt, vtx[j].pt, ORIGIN);
+    }
+
+    if (!META(i) && !META(j) && META(k)) {
+        return !Collinear(vtx[j].pt - vtx[i].pt, vtx[k].pt) ?
+                CCW(ORIGIN, vtx[j].pt - vtx[i].pt, vtx[k].pt) : CCW(vtx[i].pt, vtx[j].pt, ORIGIN);
+    }
+
+    if (!META(i) && !META(j) && !META(k)) {
+        return CCW(vtx[i].pt, vtx[j].pt, vtx[k].pt);
+    }
+
+    return CCWEx(j, k, i);
```

Comment:
the function call itself recursively. Can you prove that it works finite time?

Reply:
Because there is a symmetry between the values of the function with some of its parameters being reordered - precisely circular shift and and pairwise reordering - we can process only one case per equivalent set.